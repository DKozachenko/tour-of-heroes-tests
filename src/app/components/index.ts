export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
