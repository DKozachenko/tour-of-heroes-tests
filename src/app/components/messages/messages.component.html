<div
  *ngIf="messageService.messages.length"
  automation-id="wrapper"
  pw-automation-id="messages-wrapper"
>
  <h2 automation-id="heading" pw-automation-id="messages-heading">Messages</h2>
  <button
    type="button"
    class="clear"
    (click)="messageService.clear()"
    automation-id="clear-button"
    pw-automation-id="messages-clear-button"
  >
    Clear messages
  </button>
  <div
    *ngFor="let message of messageService.messages"
    automation-id="message"
    pw-automation-id="message"
  >
    {{ message }}
  </div>
</div>
