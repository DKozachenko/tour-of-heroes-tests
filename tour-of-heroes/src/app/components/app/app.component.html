<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
