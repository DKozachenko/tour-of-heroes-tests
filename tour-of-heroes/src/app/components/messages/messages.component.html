<div *ngIf="messageService.messages.length" automation-id="wrapper">
  <h2 automation-id="heading">Messages</h2>
  <button
    type="button"
    class="clear"
    (click)="messageService.clear()"
    automation-id="clear-button"
  >
    Clear messages
  </button>
  <div *ngFor="let message of messageService.messages" automation-id="message">
    {{ message }}
  </div>
</div>
