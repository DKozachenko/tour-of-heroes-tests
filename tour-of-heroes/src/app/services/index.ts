export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
