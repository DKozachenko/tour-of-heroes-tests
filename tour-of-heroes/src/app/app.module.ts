import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
