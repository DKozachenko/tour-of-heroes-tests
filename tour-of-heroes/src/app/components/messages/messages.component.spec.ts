import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';
import { mock, instance, when, verify } from 'ts-mockito';
import { MessageService } from '../../services';
import { MessagesComponent } from './messages.component';
import { AppModule } from '../../app.module';

class PageObject {
  private fixtureDebugElement: DebugElement;

  constructor(fixture: MockedComponentFixture<MessagesComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get divWrapper(): DebugElement {
    return this.fixtureDebugElement.query(By.css('div'));
  }

  get heading(): DebugElement {
    return this.fixtureDebugElement.query(By.css('h2'));
  }

  get clearButton(): DebugElement {
    return this.fixtureDebugElement.query(By.css('.clear'));
  }

  get messages(): DebugElement[] {
    return this.divWrapper.queryAll(By.css('div'));
  }
}

describe('MessagesComponent', () => {
  let mockMessageService: MessageService;

  beforeEach(() => {
    mockMessageService = mock(MessageService);

    return MockBuilder(MessagesComponent, AppModule).mock(
      MessageService,
      instance(mockMessageService)
    );
  });

  function createFixture(): MockedComponentFixture<MessagesComponent> {
    return MockRender(MessagesComponent);
  }

  function mockCalls(): void {
    when(mockMessageService.messages).thenReturn([]);
  }

  describe('Layout', () => {
    it('should not contain div wrapper if messages length equals 0', () => {
      mockCalls();

      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.divWrapper).toBeNull();
    });

    it('should contain div wrapper if messages length more than 0', () => {
      const mockedMessages = ['test message 1', 'test message 2'];
      when(mockMessageService.messages).thenReturn(mockedMessages);

      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.divWrapper).not.toBeNull();
      expect(pageObject.divWrapper.nativeElement).toMatchSnapshot();
    });

    it('should contain heading and clear button if messages length more than 0', () => {
      const mockedMessages = ['test message 1', 'test message 2'];
      when(mockMessageService.messages).thenReturn(mockedMessages);

      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.heading).not.toBeNull();
      expect(pageObject.heading.nativeElement).toMatchSnapshot();
      expect(pageObject.heading.nativeElement.textContent).toContain(
        'Messages'
      );
      expect(pageObject.clearButton).not.toBeNull();
      expect(pageObject.clearButton.nativeElement).toMatchSnapshot();
      expect(pageObject.clearButton.nativeElement.textContent).toContain(
        'Clear messages'
      );
    });

    it('should clear messages list if clear button has clicked', () => {
      let mockedMessages = ['test message 1', 'test message 2'];
      // Надо замкнуть кастомный массив сообщений
      when(mockMessageService.messages).thenCall(() => mockedMessages);
      when(mockMessageService.clear()).thenCall(() => (mockedMessages = []));

      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      pageObject.clearButton.triggerEventHandler('click');
      fixture.detectChanges();
      verify(mockMessageService.clear()).once();
      expect(pageObject.divWrapper).toBeNull();
    });

    it('should contain messages list if messages length more than 0', () => {
      const mockedMessages = [
        'test message 1',
        'test message 2',
        'test message 3',
      ];
      when(mockMessageService.messages).thenReturn(mockedMessages);
      const fixture = createFixture();
      const pageObject = new PageObject(fixture);
      expect(pageObject.messages).not.toHaveLength(0);

      for (let i = 0; i < mockedMessages.length; ++i) {
        const messageDiv = pageObject.messages[i];
        const mockMessage = mockedMessages[i];
        expect(messageDiv.nativeElement.textContent).toBe(mockMessage);
        expect(messageDiv.nativeElement).toMatchSnapshot();
      }
    });
  });
});
