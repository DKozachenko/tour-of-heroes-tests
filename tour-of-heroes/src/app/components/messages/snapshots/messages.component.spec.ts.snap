// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesComponent Layout should contain div wrapper if messages length more than 0 1`] = `
<div
  automation-id="wrapper"
>
  <h2
    automation-id="heading"
  >
    Messages
  </h2>
  <button
    automation-id="clear-button"
    class="clear"
    type="button"
  >
     Clear messages 
  </button>
  <div
    automation-id="message"
  >
     test message 1 
  </div>
  <div
    automation-id="message"
  >
     test message 2 
  </div>
  
</div>
`;

exports[`MessagesComponent Layout should contain heading and clear button if messages length more than 0 1`] = `
<h2
  automation-id="heading"
>
  Messages
</h2>
`;

exports[`MessagesComponent Layout should contain heading and clear button if messages length more than 0 2`] = `
<button
  automation-id="clear-button"
  class="clear"
  type="button"
>
   Clear messages 
</button>
`;

exports[`MessagesComponent Layout should contain messages list if messages length more than 0 1`] = `
<div
  automation-id="message"
>
   test message 1 
</div>
`;

exports[`MessagesComponent Layout should contain messages list if messages length more than 0 2`] = `
<div
  automation-id="message"
>
   test message 2 
</div>
`;

exports[`MessagesComponent Layout should contain messages list if messages length more than 0 3`] = `
<div
  automation-id="message"
>
   test message 3 
</div>
`;
